stm/stm.ml

```diff
@@ -2924,9 +2925,12 @@ let process_transaction ?(newtip=Stateid.fresh ())
     in
     stm_pperr_endline (fun () -> str "processed with " ++ pr_rc rc ++ str " }}}");
     VCS.print ();
+    CWarnings.set_current_loc None;
     rc
   with e ->
     let e = CErrors.push e in
+    let e = Loc.locate_if_not_already ?loc e in
+    CWarnings.set_current_loc None;
```

Comment:
IMHO this `set_current_loc` is already a very ugly hack, let's not mess with is in this patch as it is orthogonal to the original purpose.

Reply:
Warnings are printed in an asynchronous way. How to inform the warning printer of the current loc or of the current phase being interpreted.

One solution would be to have fully functional code and the current phase would be a field of the global state.

In the absence of a global state, how to avoid using a reference. Somehow, a bit better would be a `with_current_loc loc f` function but I fail to see how to do better. Do you have a suggestion?

Reply:
I agree, but this issue is orthogonal to the goal of this particular PR, so better to discuss changes to the warnings subsystem in its own issue/PR.

Reply:
OK, I think see: the `set_current_loc` could be removed once `coqloop_feed` is made aware of the phase.

For CoqIDE, is it relevant to transmit the "phase" or can we assume that there is always a loc? In particular, should we make the `slave_feeder` aware of the "phase"?

Reply:
The thing is that `CWarnings.set_current_loc` is totally broken, as soon as a command is parsed without the previous one being executed the info would be stale. IMHO, we cannot assume that there is a `loc`, it is the UI who has to handle that due to the possible asynchronous arrival of warnings. Moveover, the location in CoqIDE means little as it is relative to the current span. So IMHO, what the UI should do in the feeback handler if a message has no location is to use the `span_id` to locate the sentence and use that. This is in fact what's done in CoqIDE due to the relative nature of the locations there.

As the code stands today, initialization for CoqIDE will happen even before the CoqIDE-specific code is loaded, so I think no change should be necessary there for now.